features/largeVideo/actions.js

```diff
@@ -0,0 +1,155 @@
+            } else {
+                conference.selectParticipant(null);
+            }
+        }
+    };
+}
+
+/**
+ * Returns the most recent existing video track. It can be local or remote
+ * video.
+ *
+ * @param {(JitsiLocalTrack|JitsiRemoteTrack)[]} tracks - All current tracks.
+ * @returns {(JitsiLocalTrack|JitsiRemoteTrack|undefined)}
+ */
+function electLastVisibleVideo(tracks) {
+    let videoTrack;
+
+    // First we try to get most recent remote video track.
+    for (let i = tracks.length - 1; i >= 0; i--) {
+        if (tracks[i].isVideoTrack() && !tracks[i].isLocal()) {
+            videoTrack = tracks[i];
+            break;
+        }
+    }
+
+    // And if no remote video tracks are available, we select the local one.
+    if (!videoTrack) {
+        videoTrack = tracks.find(t => t.isLocal() && t.isVideoTrack());
+    }
+
+    return videoTrack;
+}
+
+/**
+ * Returns on stage participant.
+ *
+ * @param {Participant[]} participants - All participants.
+ * @param {(JitsiLocalTrack|JitsiRemoteTrack)[]} tracks - All tracks.
+ * @returns {(Participant|undefined)}
+ */
+function getOnStageParticipant(participants, tracks) {
+    // First get the pinned participant. If local participant is pinned, it will
+    // be shown on stage.
+    let onStageParticipant = participants.find(p => p.pinned);
+
+    // If no participant is pinned, get the dominant speaker. But local
+    // participant won't be on stage even if he is dominant speaker.
+    if (!onStageParticipant) {
+        onStageParticipant = participants.find(p => p.speaking && !p.local);
+    }
+
+    // If no participant is pinned and no dominant speaker, just get the
+    // participant with last visible video track. This may turn out ot be local
+    // participant.
+    if (!onStageParticipant) {
+        let videoTrack = electLastVisibleVideo(tracks);
+        onStageParticipant = getParticipantByVideoTrack(
+            videoTrack, participants);
+    }
+
+    return onStageParticipant;
+}
+
+/**
+ * Returns participant corresponding to video stream.
+ *
+ * @param {JitsiLocalTrack|JitsiRemoteTrack} track - Current video track.
+ * @param {Participant[]} participants - List of all participants.
+ * @returns {(Participant|undefined)}
+ */
+function getParticipantByVideoTrack(track, participants) {
+    if (!track) {
+        return;
+    }
+
+    return track.isLocal()
+        ? participants.find(p => p.local)
+        : participants.find(p => p.id === track.getParticipantId());
+}
+
+/**
+ * Returns video track for a specified participant.
+ *
+ * @param {Participant|undefined} participant - Participant object.
+ * @param {(JitsiLocalTrack|JitsiRemoteTrack)[]} tracks - List of all tracks.
+ * @returns {(JitsiLocalTrack|JitsiRemoteTrack|undefined)}
+ */
+function getVideoTrack(participant, tracks) {
+    if (!participant) {
+        return;
+    }
+
+    return tracks.find(t => {
+        return t.isVideoTrack() &&
```

Comment:
Another duplication that I've removed during my review are the 3 lines bellow. They represent the logic of isParticipantTrack(participant, track) defined as _isParticipantTrack in FilmStrip.js.

While I'd like to note duplications, I'm writing, @tsareg, mostly because I left both isParticipantTrack & getVideoTrack in actions.js files even though they are not actions. I'd like us to come up with good locations for them in the source tree of the project but I want to merge this PR ASAP. So I'm communicating with you here so that you & I can give it a few thoughts & have a short talk tomorrow. Thank you!